shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Rules/Utxo.hs

```diff
@@ -0,0 +1,376 @@
+    State (UTXO era) ~ Shelley.UTxOState era,
+    Signal (UTXO era) ~ Tx era,
+    PredicateFailure (UTXO era) ~ UtxoPredicateFailure era,
+    HasField "certs" (Core.TxBody era) (StrictSeq (DCert era)),
+    HasField "inputs" (Core.TxBody era) (Set (TxIn era)),
+    HasField "forge" (Core.TxBody era) (Core.Value era),
+    HasField "outputs" (Core.TxBody era) (StrictSeq (TxOut era)),
+    HasField "wdrls" (Core.TxBody era) (Wdrl era),
+    HasField "txfee" (Core.TxBody era) Coin,
+    HasField "vldt" (Core.TxBody era) ValidityInterval,
+    HasField "update" (Core.TxBody era) (StrictMaybe (Update era))
+  ) =>
+  TransitionRule (UTXO era)
+utxoTransition = do
+  TRC (Shelley.UtxoEnv slot pp stakepools genDelegs, u, tx) <- judgmentContext
+  let Shelley.UTxOState utxo deposits' fees ppup = u
+  let txb = _body tx
+
+  inInterval slot (getField @"vldt" txb)
+    ?! OutsideValidityIntervalUTxO (getField @"vldt" txb) slot
+
+  txins txb /= Set.empty ?! InputSetEmptyUTxO
+
+  let minFee = Shelley.minfee pp tx
+      txFee = getField @"txfee" txb
+  minFee <= txFee ?! FeeTooSmallUTxO minFee txFee
+
+  eval (txins txb ⊆ dom utxo)
+    ?! BadInputsUTxO (txins txb `Set.difference` eval (dom utxo))
+
+  ni <- liftSTS $ asks networkId
+  let addrsWrongNetwork =
+        filter
+          (\a -> getNetwork a /= ni)
+          (fmap (\(TxOut a _) -> a) $ toList $ getField @"outputs" txb)
+  null addrsWrongNetwork ?! WrongNetwork ni (Set.fromList addrsWrongNetwork)
+  let wdrlsWrongNetwork =
+        filter
+          (\a -> getRwdNetwork a /= ni)
+          (Map.keys . unWdrl . getField @"wdrls" $ txb)
+  null wdrlsWrongNetwork
+    ?! WrongNetworkWithdrawal
+      ni
+      (Set.fromList wdrlsWrongNetwork)
+
+  let consumed_ = consumed pp utxo txb
+      produced_ = Shelley.produced pp stakepools txb
+  consumed_ == produced_ ?! ValueNotConservedUTxO consumed_ produced_
+
+  -- process Protocol Parameter Update Proposals
+  ppup' <- trans @(PPUP era) $ TRC (PPUPEnv slot pp genDelegs, ppup, txup tx)
+
+  -- Check that the forge field does not try to forge ADA. This is equivalent to
+  -- the check `adaPolicy ∉ supp forge tx` in the spec.
+  Val.coin (getField @"forge" txb) == Val.zero ?! TriesToForgeADA
+
+  let outputs = Map.elems $ unUTxO (txouts txb)
+      minUTxOValue = _minUTxOValue pp
+      outputsTooSmall =
+        [ out
+          | out@(TxOut _ c) <- outputs,
+            Val.pointwise
+              (<)
+              c
+              (Val.inject $ Val.scaledMinDeposit c minUTxOValue)
+        ]
+  null outputsTooSmall ?! OutputTooSmallUTxO outputsTooSmall
+
+  -- Bootstrap (i.e. Byron) addresses have variable sized attributes in them.
+  -- It is important to limit their overall size.
+  let outputsAttrsTooBig =
+        [ out
+          | out@(TxOut (AddrBootstrap addr) _) <- outputs,
+            bootstrapAddressAttrsSize addr > 64
+        ]
+  null outputsAttrsTooBig ?! OutputBootAddrAttrsTooBig outputsAttrsTooBig
```

Comment:
do we need to check the size of the AssetID in the same way (it should have length <=32)
or is that getting enforced in the CBOR or CDDL?

Reply:
I think Alex was going to ask you about this. Is it defined in the spec somewhere?